yarn-project/simulator/src/avm/journal/nullifiers.ts

```diff
@@ -41,10 +40,12 @@ export class Nullifiers {
     nullifier: Fr,
   ): Promise<[/*exists=*/ boolean, /*isPending=*/ boolean, /*leafIndex=*/ Fr]> {
     // First check this cache
-    let existsAsPending = this.cache.exists(storageAddress, nullifier);
-    // Then check parent's cache
-    if (!existsAsPending && this.parentCache) {
-      existsAsPending = this.parentCache?.exists(storageAddress, nullifier);
+    const existsAsPending = this.cache.exists(storageAddress, nullifier);
+    // Then try parent's nullifier cache
+    if (!existsAsPending && this.parent) {
+      // Note: this will recurse to grandparent/etc until a cache-hit is encountered.
+      // Otherwise it will fall back to host state in the top-most [grand]parent.
+      return await this.parent.checkExists(storageAddress, nullifier);
```

Comment:
This surprised me! The "return"!

I was still thinking that we'd continue if not found in the parent, which is ok, but would end up doing a host-check for every member of the chain.

I see this is solved by returning here, but it's a bit surprising, since it "only" works because we will not reach this return in the base case (the root), and we will continue and check the host.

Maybe add a bit more comment? (same in storage). Maybe it's just me and this pattern is common?

Reply:
I'll create a private helper function that just checks caches recursively (no host check).